native/game_backend/src/player.rs

```diff
@@ -93,22 +97,21 @@ impl Player {
     }
 
     pub fn add_action(&mut self, action: Action, duration_ms: u64) {
-        if !self.actions.contains(&action) {
-            self.next_actions.push(action);
-            self.action_duration_ms += duration_ms;
-        }
+        self.next_actions.push(ActionTracker {
+            action,
+            duration: duration_ms,
+        });
     }
 
-    pub fn update_actions(&mut self) {
-        if !self.is_executing_action() {
-            self.next_actions
-                .retain(|action| matches!(action, Action::Moving));
-        }
-        self.actions = self.next_actions.clone();
-    }
+    pub fn update_actions(&mut self, elapsed_time_ms: u64) {
+        if let Some(action_tracker) = self.next_actions.first_mut() {
+            action_tracker.duration = action_tracker.duration.saturating_sub(elapsed_time_ms);
 
-    fn is_executing_action(&mut self) -> bool {
-        self.action_duration_ms != 0
+            if action_tracker.duration == 0 {
+                self.next_actions.remove(0);
+            }
+        }
+        self.action = self.next_actions.clone();
```

Comment:
Maybe having 2 vectors is unnecessary, because we have the actions updated all the time. But is not a blocker. We can fix it in another PR.